Deploy codabench using kubernetes

pauloguilhermepp opened the pull request:
#  Deploy Codabench Using Kubernetes
This PR aims to make Codabench deployable using Kubernetes.

# Main Changes
To deploy Codabench using Kubernetes, we:
* Added Helm charts for all core components.
* Updated Dockerfiles to improve compatibility with Kubernetes deployments.

# Issues this PR resolves
* Closes #1933

# How to Test it
1. Prerequisites
    * [Kubectl](https://kubernetes.io/docs/reference/kubectl/)
    * [Helm](https://helm.sh/)
    * Access to a kubernetes cluster (like [minikube](https://minikube.sigs.k8s.io/docs/start/?arch=%2Flinux%2Fx86-64%2Fstable%2Fbinary+download) or [kind](https://kind.sigs.k8s.io/))
2. Start your cluster (eg, `minikube start`)
3. Fill in `values.yaml` properly
4. Deploy the Helm chart
```
helm install <project_name> . -n <namespace> --create-namespace -f <path_to_values>
```
5. Check the status of your pods
```
kubectl get pods
```

# Observations

* Tests mainly made at [1.19](https://github.com/codalab/codabench/tree/1.19.0). 

* To expose the codabench UI, you will have to install an [ingress controller](https://kubernetes.io/docs/concepts/services-networking/ingress/) and create an Ingress.

* This PR does not include a template for Minio yet. For now, testing requires an external S3 instance. We can also add the Minio chart if needed.

* The helm chart passes secret values directly via environment variables, which is not the best practice, as it will make them visible in deployment tools like ArgoCD. If you want, we can update this PR (or open a future one) with this change.

* Inside values.yaml, there is a section (env) with environment variables that are currently passed to multiple pods. This is similar to the .env setup in docker-compose, but as we sometimes had a hard time debugging which variable is used where, we would like to separate those in a later PR.

* This PR is making Codabench deployable using Kubernetes, but it is not changing how the compute-worker is running the user submissions. We plan to open another PR in the future, updating the compute-worker to run submissions using Kubernetes. We plan to make it configurable and using docker as default so that it will not break any current deployments. Please note that for the current way the compute worker works, the Kubernetes Pods need to mount a volume from the host node, to mount the Docker socket from the node, and run privileged Docker containers, which is not the best practice. This will mess with the storage on the node and will mess with the scheduling, as containers are directly created, bypassing the Kubernetes scheduler. Therefore, we will open another PR.

* The PR also contains the Dockerfiles that we are using. Ideally, we should see if they are still compatible with deploying Codabench using the original docker-compose setup and keep a single Dockerfile for each component. The images were updated because, ideally, we don't want to mount the code from a PersistentVolume or the host node but rather have the image contain the code.

# Checklist
- [x] Code review by me 
- [x] Hand tested by me 
- [x] I'm proud of my work
- [x] Code review by reviewer
- [x] Hand tested by reviewer
- [x] CircleCi tests are passing
- [x] Ready to merge

hahahannes commented on charts/Chart.yaml:
```diff
@@ -0,0 +1,34 @@
+dependencies:
+  - name: rabbitmq
+    version: "14.7.0"
+    repository: "oci://registry.cern.ch/kubeflow/charts"
```
We will replace this one with the original upstream chart https://hub.docker.com/r/bitnamicharts/rabbitmq. Same for redis below.

hahahannes commented on charts/dockerfiles/Dockerfile:
```diff
@@ -0,0 +1,43 @@
+RUN poetry install
+
+# Copy the rest of the application code
+COPY . /app
```
Regarding the discussion about the Dockerfiles. We saw that the django component was missing the code (https://github.com/codalab/codabench/blob/develop/Dockerfile) also in the latest develop branch. Therefor we added node builder image here as well as the copying of the code.

pauloguilhermepp commented:
Hi everyone.

We rebased the branch and also updated the Django image so it is self-contained.

We successfully installed the Chart using this updated image.

pauloguilhermepp commented:
Hi everyone,

With the conflicts fixed in the PR, it should now be possible to merge it. 

[Containerfile](https://github.com/pauloguilhermepp/codabench/blob/develop/packaging/container/Containerfile) is the only change that also affects the docker-compose setup. As we mentioned, the images must be self-contained to be deployed using Kubernetes. This means that all code used by the image must be contained within it.

The new Containerfile combines the [old Containerfile](https://github.com/codalab/codabench/blob/develop/packaging/container/Containerfile) with [Container.builder](https://github.com/codalab/codabench/blob/develop/packaging/container/Containerfile.builder) in such a way that all Django dependencies will be present in the final image. If you consider that this is incompatible with the docker-compose deployment, we can add a new Containerfile with such changes. 

Please, feel free to share your feedback.

Didayolo commented:
Thank you for your work!

Post-it: we'll need to rebase and resolve conflicts again after merging #2223.

---

EDIT: I thought this PR modified `compute_worker.py`. It's all good now, and `fix-worker-logs` is rebased.
